source/extensions/tracers/datadog/tracer.cc

```diff
@@ -0,0 +1,126 @@
+    config.logger->log_error(error->with_prefix(prefix));
+    return std::make_shared<Tracer::ThreadLocalTracer>();
+  }
+
+  return std::make_shared<Tracer::ThreadLocalTracer>(*maybe_config);
+}
+
+} // namespace
+
+Tracer::ThreadLocalTracer::ThreadLocalTracer(const datadog::tracing::FinalizedTracerConfig& config)
+    : tracer(config) {}
+
+Tracer::Tracer(const std::string& collector_cluster, const std::string& collector_reference_host,
+               const datadog::tracing::TracerConfig& config,
+               Upstream::ClusterManager& cluster_manager, Stats::Scope& scope,
+               ThreadLocal::SlotAllocator& thread_local_slot_allocator)
+    : tracer_stats_(makeTracerStats(scope)),
+      thread_local_slot_(thread_local_slot_allocator.allocateSlot()) {
+  const bool allow_added_via_api = true;
+  Config::Utility::checkCluster("envoy.tracers.datadog", collector_cluster, cluster_manager,
+                                allow_added_via_api);
+
+  thread_local_slot_->set([&logger = ENVOY_LOGGER(), collector_cluster, collector_reference_host,
+                           config, &tracer_stats = tracer_stats_,
+                           &cluster_manager](Event::Dispatcher& dispatcher) {
+    return makeThreadLocalTracer(config, cluster_manager, collector_cluster,
+                                 collector_reference_host, tracer_stats, dispatcher, logger);
+  });
+}
+
+// Tracer::TracingDriver
+
+Tracing::SpanPtr Tracer::startSpan(const Tracing::Config&, Tracing::TraceContext& trace_context,
+                                   const std::string& operation_name, SystemTime start_time,
+                                   const Tracing::Decision tracing_decision) {
+  auto& thread_local_tracer = thread_local_slot_->getTyped<ThreadLocalTracer>();
+  if (!thread_local_tracer.tracer) {
+    return std::make_unique<Tracing::NullSpan>();
+  }
+
+  // The OpenTracing implementation ignored the `Tracing::Config` argument,
+  // so we will as well.
+  datadog::tracing::SpanConfig span_config;
+  span_config.name = operation_name;
+  span_config.start = estimateTime(start_time);
+
+  datadog::tracing::Tracer& tracer = *thread_local_tracer.tracer;
+  TraceContextReader reader{trace_context};
+  auto maybe_span = tracer.extract_span(reader, span_config);
+  if (auto* error = maybe_span.if_error()) {
+    // We didn't extract a span. Either there's no span to extract, or an
+    // error occurred during extraction.
+    //
+    // Either way, we're going to create a new root span, but if an error
+    // occurred we're going to log the error.
+    if (error->code != datadog::tracing::Error::NO_SPAN_TO_EXTRACT) {
+      ENVOY_LOG(
+          error,
+          "Unable to extract span context. Creating a new trace instead. Error [error {}]: {}",
+          int(error->code), error->message);
+    }
+
+    maybe_span = tracer.create_span(span_config);
+  }
+
+  assert(maybe_span);
```

Comment:
ASSERT or RELEASE_ASSERT per Envoy style?

RELEASE_ASSERT would be more like assert() (can abort production binaries)